compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
-
-            // Other `pub` items inherit levels from parents.
-            ast::ItemKind::ExternCrate(..)
-            | ast::ItemKind::Use(..)
-            | ast::ItemKind::Static(..)
-            | ast::ItemKind::Const(..)
-            | ast::ItemKind::Fn(..)
-            | ast::ItemKind::Mod(..)
-            | ast::ItemKind::GlobalAsm(..)
-            | ast::ItemKind::TyAlias(..)
-            | ast::ItemKind::Enum(..)
-            | ast::ItemKind::Struct(..)
-            | ast::ItemKind::Union(..)
-            | ast::ItemKind::Trait(..)
-            | ast::ItemKind::TraitAlias(..)
-            | ast::ItemKind::MacroDef(..) => {
-                if item.vis.kind.is_pub() {
-                    self.prev_level
-                } else {
-                    None
-                }
-            }
-
             // Should be unreachable at this stage
             ast::ItemKind::MacCall(..) => panic!(
                 "ast::ItemKind::MacCall encountered, this should not anymore appear at this stage"
             ),
-        };
 
-        let access_level = self.set_access_level(item.id, inherited_item_level);
+            // Foreign modules inherit level from parents.
+            ast::ItemKind::ForeignMod(..) => {
+                self.set_access_level(item.id, prev_level);
+            }
 
-        // Set access level of nested items.
-        // If it's a mod, also make the visitor walk all of its items
-        match item.kind {
-            ast::ItemKind::Mod(..) => {
-                if access_level.is_some() {
-                    self.set_exports_access_level(self.r.local_def_id(item.id));
+            // Only exported `macro_rules!` items are public, but they always are
+            ast::ItemKind::MacroDef(ref macro_def) if macro_def.macro_rules => {
+                if item.attrs.iter().any(|attr| attr.has_name(sym::macro_export)) {
+                    self.set_access_level(item.id, Some(AccessLevel::Public));
                 }
-
-                let orig_level = std::mem::replace(&mut self.prev_level, access_level);
-                visit::walk_item(self, item);
-                self.prev_level = orig_level;
             }
 
-            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
-                for nested in items {
-                    if nested.vis.kind.is_pub() {
-                        self.set_access_level(nested.id, access_level);
-                    }
+            ast::ItemKind::Mod(..) => {
+                let current_level = self.r.access_levels.map.get(&def_id);
+                if current_level.is_some() {
+                    self.set_bindings_access_level(def_id);
                 }
+                visit::walk_item(self, item);
             }
+
             ast::ItemKind::Enum(EnumDef { ref variants }, _) => {
                 for variant in variants {
                     let variant_level = self.set_access_level(variant.id, access_level);
```

Comment:
Now you are (incorrectly) setting the level for variants, but not for variant constructors.
Could you use `set_bindings_access_level` here as I suggested in https://github.com/rust-lang/rust/pull/100147/files#r944705524?